k9mail/src/main/java/com/fsck/k9/notification/WearNotifications.java

```diff
@@ -0,0 +1,205 @@
+    }
+
+    private void addArchiveAllAction(WearableExtender wearableExtender, NotificationsHolder notificationsHolder) {
+        int icon = R.drawable.ic_action_archive_dark;
+        String title = context.getString(R.string.notification_action_archive_all);
+
+        Account account = notificationsHolder.getAccount();
+        ArrayList<MessageReference> messageReferences = notificationsHolder.getAllMessageReferences();
+        int notificationId = NotificationIds.getNewMailNotificationId(account);
+        PendingIntent action = actionCreator.createArchiveAllPendingIntent(account, messageReferences, notificationId);
+
+        NotificationCompat.Action archiveAction = new NotificationCompat.Action.Builder(icon, title, action).build();
+        wearableExtender.addAction(archiveAction);
+    }
+
+    private void addActions(Builder builder, Account account, NotificationHolder holder) {
+        NotificationCompat.WearableExtender wearableExtender = new NotificationCompat.WearableExtender();
+
+        addReplyAction(wearableExtender, holder);
+        addMarkAsReadAction(wearableExtender, holder);
+
+        if (isDeleteActionAvailableForWear()) {
+            addDeleteAction(wearableExtender, holder);
+        }
+
+        if (isArchiveActionAvailableForWear(account)) {
+            addArchiveAction(wearableExtender, holder);
+        }
+
+        if (isSpamActionAvailableForWear(account)) {
+            addMarkAsSpamAction(wearableExtender, holder);
+        }
+
+        builder.extend(wearableExtender);
+    }
+
+    private void addReplyAction(WearableExtender wearableExtender, NotificationHolder holder) {
+        int icon = R.drawable.ic_action_single_message_options_dark;
+        String title = context.getString(R.string.notification_action_reply);
+
+        MessageReference messageReference = holder.content.messageReference;
+        int notificationId = holder.notificationId;
+        PendingIntent action = actionCreator.createReplyPendingIntent(messageReference, notificationId);
+
+        NotificationCompat.Action replyAction = new NotificationCompat.Action.Builder(icon, title, action).build();
+        wearableExtender.addAction(replyAction);
+    }
+
+    private void addMarkAsReadAction(WearableExtender wearableExtender, NotificationHolder holder) {
+        int icon = R.drawable.ic_action_mark_as_read_dark;
+        String title = context.getString(R.string.notification_action_mark_as_read);
+
+
+        MessageReference messageReference = holder.content.messageReference;
+        int notificationId = holder.notificationId;
+        PendingIntent action = actionCreator.createMarkMessageAsReadPendingIntent(messageReference, notificationId);
+
+        NotificationCompat.Action markAsReadAction = new NotificationCompat.Action.Builder(icon, title, action).build();
+        wearableExtender.addAction(markAsReadAction);
+    }
+
+    private void addDeleteAction(WearableExtender wearableExtender, NotificationHolder holder) {
```

Comment:
The code itself should work fine when the condition (`isDeleteActionAvailable()`) is not satisfied. It's just that the behavior the user observes might be undesired.

I believe my original version of that code was similar to what you suggested. I then extracted the checks to the call site so the methods to add actions all have the same arguments and follow the same structure. Given that the methods are private to the class I'm okay with not checking the "preconditions" inside the method.